route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -43,35 +42,47 @@ namespace route_following_plugin
         plugin_discovery_msg_.capability = "strategic_plan/plan_maneuvers";
         pose_sub_ = nh_->subscribe("current_pose", 1, &RouteFollowingPlugin::pose_cb, this);
         twist_sub_ = nh_->subscribe("current_velocity", 1, &RouteFollowingPlugin::twist_cd, this);
-        
-        pnh_->param<double>("minimal_maneuver_duration", mvr_duration_, 16.0);
-        pnh2_->param<double>("config_speed_limit",config_limit);
-        pnh_->param<double>("buffer_time_lanechange",buffer_lanechange_time_);
-        pnh2_->param<double>("guidance/route_end_jerk", jerk_, 1.0);
-        pnh2_->param<std::string>("guidance/lane_change_plugin", lane_change_plugin_);
+
         wml_.reset(new carma_wm::WMListener());
         // set world model point form wm listener
         wm_ = wml_->getWorldModel();
-
+        //set a route callback to update route and calculate maneuver
+        wml_->setRouteCallback(route_set);
+        if(route_set)    
+        {
+            route_cb();
+        }
+        
         discovery_pub_timer_ = pnh_->createTimer(
-            ros::Duration(ros::Rate(10.0)),
-            [this](const auto&) { plugin_discovery_pub_.publish(plugin_discovery_msg_); });
+        ros::Duration(ros::Rate(10.0)),
+        [this](const auto&) { plugin_discovery_pub_.publish(plugin_discovery_msg_); });
     }
+
     void RouteFollowingPlugin::run()
     {
         initialize();
         ros::CARMANodeHandle::spin();
     }
-    bool RouteFollowingPlugin::plan_maneuver_cb(cav_srvs::PlanManeuversRequest &req, cav_srvs::PlanManeuversResponse &resp)
-    {        
+
+    void RouteFollowingPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+    }
+    void RouteFollowingPlugin::twist_cd(const geometry_msgs::TwistStampedConstPtr& msg)
+    {
+        current_speed_ = msg->twist.linear.x;
+    }
+
+    void RouteFollowingPlugin::route_cb(){
+        //This function calculates the maneuver every time a route is set
+        //Go through route - identify lane changes and fill in the spaces with lane following
         lanelet::BasicPoint2d current_loc(pose_msg_.pose.position.x, pose_msg_.pose.position.y);
-        auto current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, current_loc, 10);       
+        auto current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, current_loc, 10); 
```

Comment:
nearest_lanelets would be a better name for this vector since only a few of them may actually be a lanelet we are on top of.

Reply:
Agree with Mike.  Also a comment quickly explaining what the 10 argument means (number of points? search radius?), since I spent several minutes looking for the proper lanelet header file that declares this function (there are at least three LaneletMap.h files!!!).  And since it is number of neighbors, why is 10 the desired number?

Reply:
@TonysCousin findNearest is a k-nearest-neighbors method so 10 means the number of nearest lanelets to return. They are returned in distance order

Reply:
Good explanation, Mike, but it's nice to have it written in the code so I don't have to go dig for 10 minutes next year to refresh my memory.